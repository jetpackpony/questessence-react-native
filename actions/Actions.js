import Database from '../database/Database';
import PurchaseAPI from '../purchaseApi/PurchaseAPI';
import Config from 'react-native-config';

export const QuestStates = {
  PURCHASED: 'PURCHASED',
  DOWNLOADING: 'DOWNLOADING',
  NOT_STARTED: 'NOT_STARTED',
  IN_PROGRESS: 'IN_PROGRESS',
  COMPLETED: 'COMPLETED'
};

export const QuestionStates = {
  UNANSWERED: 'UNANSWERED',
  CORRECT: 'CORRECT',
  INCORRECT: 'INCORRECT',
  SHOW_ANSWER: 'SHOW_ANSWER'
};

export function startQuest(questId) {
  return { type: 'START_QUEST', questId };
};

export function answerQuestion(questId, questionId, answer) {
  return { type: 'ANSWER_QUESTION', questId, questionId, answer };
};

export function showCorrectAnswer(questId) {
  return { type: 'SHOW_CORRECT_ANSWER', questId };
};

export function goToNextQuestion(questId) {
  return { type: 'GOTO_NEXT_QUESTION', questId };
};

export function updateQuestList(quests) {
  return { type: 'UPDATE_QUEST_LIST', quests };
};

export function purchaseQuest(questId, productId) {
  return (dispatch) => {
    dispatch({ type: 'PURCHASE_QUEST_START', questId, productId });

    if (Config.USE_DUMMY_GOOGLE_PLAY_PODUCTS === "true") {
      productId = "android.test.purchased";
    }
    PurchaseAPI.purchase(productId)
      .then((res) => {
        console.log('You purchased: ', res);
        return PurchaseAPI.consume(productId);
      })
      .then(() => {
        dispatch({ type: 'PURCHASE_QUEST_SUCCESS', questId, productId });
      })
      .catch((err) => {
        console.log(err);
      });
  };
}

export function downloadQuest(questId) {
  return (dispatch) => {
    dispatch({ type: 'DOWNLOADING_QUEST_START', questId });

    Database
      .downloadQuestions(questId)
      .then((questions) => {
        dispatch({ type: 'DOWNLOADING_QUEST_SUCCESS', questId, questions });
      });
  };
};

export function deleteQuest(questId) {
  return { type: 'DELETE_QUEST', questId };
}
