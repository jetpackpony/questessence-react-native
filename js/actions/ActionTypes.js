export default {
  START_QUEST: 'START_QUEST',
  ANSWER_QUESTION: 'ANSWER_QUESTION',
  SHOW_CORRECT_ANSWER: 'SHOW_CORRECT_ANSWER',
  GOTO_NEXT_QUESTION: 'GOTO_NEXT_QUESTION',
  RESTART_QUEST: 'RESTART_QUEST',

  UPDATE_QUEST_LIST: 'UPDATE_QUEST_LIST',

  DOWNLOADING_QUEST_START: 'DOWNLOADING_QUEST_START',
  DOWNLOADING_QUEST_SUCCESS: 'DOWNLOADING_QUEST_SUCCESS',

  LOGIN_START: 'LOGIN_START',
  LOGIN_SUCCESS: 'LOGIN_SUCCESS',
  LOGOUT: 'LOGOUT',

  SHOW_LOGIN_MODAL: 'SHOW_LOGIN_MODAL',
  HIDE_LOGIN_MODAL: 'HIDE_LOGIN_MODAL',

  SYNC_PROGRESS_SUCCESS: 'SYNC_PROGRESS_SUCCESS',

  DELETE_QUEST: 'DELETE_QUEST'
};
