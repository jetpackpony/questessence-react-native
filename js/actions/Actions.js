import types from './ActionTypes';

export const ActionTypes = types;

export {
  QuestStates, DownloadStates, QuestionStates
} from './constants';

export {
  startQuest, answerQuestion,
  showCorrectAnswer, goToNextQuestion
} from './QuestProgressActions';

export {
  downloadQuest, downloadQuestStart,
  downloadQuestSuccess, updateQuestList
} from './DownloadActions';

export {
  loginFacebook, loginFirebaseFacebook,
  restoreLogin, logout,
  loginStart, loginSuccess,
  logoutSuccess, showLoginModal,
  hideLoginModal
} from './LoginActions';

export {
  syncProgress, syncProgressSuccess
} from './SyncActions';



export function deleteQuest(questId) {
  return { type: ActionTypes.DELETE_QUEST, questId, updatingProgress: true };
}

