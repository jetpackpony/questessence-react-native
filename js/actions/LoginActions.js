import types from './ActionTypes';
import { AccessToken } from 'react-native-fbsdk';
import firebase from '../database/firebase';

import { syncProgress } from './SyncActions';

export function loginFacebook(error, result) {
  return (dispatch) => {
    if (error) {
      console.log("login has error: ", error);
    } else if (result.isCancelled) {
      console.log("login is cancelled.", result);
    } else {
      dispatch(loginFirebaseFacebook());
    }
  };
}

export function loginFirebaseFacebook() {
  return (dispatch) => {
    AccessToken.getCurrentAccessToken()
      .then((userData) => {
        let token = userData.accessToken.toString();
        let credential = firebase.auth.FacebookAuthProvider.credential(token);
        firebase.auth().signInWithCredential(credential)
          .then(() => {
            dispatch(loginSuccess(firebase.auth().currentUser));
            dispatch(syncProgress());
          })
          .catch(() => {
            console.log('FIREBASE ERROR: ', error);
          });
      })
      .catch((error) => {
        // If there is not token, logout from firebase
        dispatch(logout());
      });
  };
}

export function restoreLogin() {
  return (dispatch) => {
    dispatch(loginFirebaseFacebook());
  };
}

export function logout() {
  return (dispatch) => {
    firebase.auth().signOut()
      .then(() => dispatch(logoutSuccess()));
  };
}

export function loginStart() {
  return { type: types.LOGIN_START };
}

export function loginSuccess(user) {
  return { type: types.LOGIN_SUCCESS, user };
}

export function logoutSuccess() {
  return { type: types.LOGOUT };
}

export function showLoginModal() {
  return { type: types.SHOW_LOGIN_MODAL };
}

export function hideLoginModal() {
  return { type: types.HIDE_LOGIN_MODAL };
}
