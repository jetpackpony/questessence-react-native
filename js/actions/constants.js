export const QuestStates = {
  PURCHASED: 'PURCHASED',
  NOT_STARTED: 'NOT_STARTED',
  IN_PROGRESS: 'IN_PROGRESS',
  COMPLETED: 'COMPLETED'
};

export const DownloadStates = {
  NOT_DOWNLOADED: 'NOT_DOWNLOADED',
  DOWNLOADING: 'DOWNLOADING',
  DOWNLOADED: 'DOWNLOADED'
};

export const QuestionStates = {
  UNANSWERED: 'UNANSWERED',
  CORRECT: 'CORRECT',
  INCORRECT: 'INCORRECT',
  SHOW_ANSWER: 'SHOW_ANSWER'
};
