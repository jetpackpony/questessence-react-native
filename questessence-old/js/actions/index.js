import types from './ActionTypes';

export const ActionTypes = types;

export {
  startQuest, answerQuestion, showCorrectAnswer,
  goToNextQuestion, restartQuest
} from './QuestProgressActions';

export {
  downloadQuest, downloadQuestStart,
  downloadQuestSuccess, updateQuestList,
  deleteQuest
} from './DownloadActions';

export {
  loginFacebook, loginFirebaseFacebook,
  restoreLogin, logout,
  loginStart, loginSuccess,
  logoutSuccess, showLoginModal,
  hideLoginModal
} from './LoginActions';

export {
  syncProgress, syncProgressSuccess
} from './SyncActions';

