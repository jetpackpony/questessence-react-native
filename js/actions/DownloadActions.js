import Database from '../database/Database';
import types from './ActionTypes';
import { showLoginModal } from './Actions';

export function updateQuestList(quests) {
  return { type: types.UPDATE_QUEST_LIST, quests };
};

export function downloadQuest(questId) {
  return (dispatch) => {
    dispatch(downloadQuestStart(questId));

    Database
      .downloadQuestions(questId)
      .then((questions) => {
        dispatch(downloadQuestSuccess(questId, questions));
        dispatch(showLoginModal());
      });
  };
};

export function downloadQuestStart(questId) {
  return { type: types.DOWNLOADING_QUEST_START, questId };
}

export function downloadQuestSuccess(questId, questions) {
  return { type: types.DOWNLOADING_QUEST_SUCCESS, questId, questions, updatingProgress: true  };
}

