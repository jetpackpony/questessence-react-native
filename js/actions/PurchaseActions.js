import PurchaseAPI from '../purchaseApi/PurchaseAPI';
import { useDummyGoogleProductID } from '../config';
import types from './ActionTypes';

import { showLoginModal } from './Actions';

export function purchaseQuest(questId, productId) {
  return (dispatch) => {
    dispatch(purchaseQuestStart(questId, productId));

    if (useDummyGoogleProductID) {
      productId = "android.test.purchased";
    }
    PurchaseAPI.purchase(productId)
      .then((res) => {
        console.log('You purchased: ', res);
        if (useDummyGoogleProductID) {
          return PurchaseAPI.consume(productId);
        }
      })
      .then(() => {
        dispatch(purchaseQuestSuccess(questId, productId));
        dispatch(showLoginModal());
      })
      .catch((err) => {
        console.log(err);
      });
  };
}

export function purchaseQuestStart(questId, productId) {
  return { type: types.PURCHASE_QUEST_START, questId, productId };
}

export function purchaseQuestSuccess(questId, productId) {
  return { type: types.PURCHASE_QUEST_SUCCESS, questId, productId, updatingProgress: true };
}

