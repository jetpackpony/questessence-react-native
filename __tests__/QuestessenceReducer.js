import {
  QuestessenceReducer,
  initialState
} from '../reducers/QuestessenceReducer';
import { startQuest, answerQuestion } from '../actions/Actions.js';
import { QuestStates, QuestionStates } from '../actions/Actions';

const addDummyQuest = (inState) => {
  let state = { ...inState };
  state.questData.quests.push({
    id: 0,
    title: 'Test',
    cover: 'http://...',
    desc: 'Description'
  });
  state.questData.questions.quest0 = [{
    media: {
      type: 'image',
      uri: 'https://'
    },
    desc: 'На этом месте будет описание ',
    questionText: 'Сколько планет в солнечной системе?',
    answer: '8',
    answerDesc: 'Потому что Плутон больше не планета'
  },{
    media: {
      type: 'image',
      uri: 'https://'
    },
    desc: 'На этом месте будет описание вопроса',
    questionText: 'Сколько часов в сутках?',
    answer: '24',
    answerDesc: 'Хотелось бы больше, конечно'
  }];
  state.questProgress.quest0 = {
    questState: QuestStates.IN_PROGRESS,
    currentQuestion: 0,
    currentAnswer: '',
    currentQuestionState: QuestionStates.UNANSWERED
  }
  return state;
};

it('starts a quest', () => {
  expect(
    QuestessenceReducer(initialState, startQuest(0))
  ).toMatchSnapshot();
});

it('sets the question state if the answer is correct', () => {
  let state = addDummyQuest(initialState);
  state.questData.questions.quest0[0].answer = '8';
  expect(
    QuestessenceReducer(state, answerQuestion(0, 0, '8'))
  ).toMatchSnapshot();
});

it('sets the question state if the answer is incorrect', () => {
  let state = addDummyQuest(initialState);
  state.questData.questions.quest0[1].answer = '24';
  expect(
    QuestessenceReducer(state, answerQuestion(0, 1, '15'))
  ).toMatchSnapshot();
});
