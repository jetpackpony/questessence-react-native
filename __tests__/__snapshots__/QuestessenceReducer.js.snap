// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sets the question state if the answer is correct 1`] = `
Object {
  "questData": Object {
    "questions": Object {
      "quest0": Array [
        Object {
          "answer": "8",
          "answerDesc": "Потому что Плутон больше не планета",
          "desc": "На этом месте будет описание ",
          "media": Object {
            "type": "image",
            "uri": "https://",
          },
          "questionText": "Сколько планет в солнечной системе?",
        },
        Object {
          "answer": "24",
          "answerDesc": "Хотелось бы больше, конечно",
          "desc": "На этом месте будет описание вопроса",
          "media": Object {
            "type": "image",
            "uri": "https://",
          },
          "questionText": "Сколько часов в сутках?",
        },
      ],
    },
    "quests": Array [
      Object {
        "cover": "http://...",
        "desc": "Description",
        "id": 0,
        "title": "Test",
      },
    ],
  },
  "questProgress": Object {
    "quest0": Object {
      "currentAnswer": "8",
      "currentQuestion": 0,
      "currentQuestionState": "CORRECT",
      "questState": "IN_PROGRESS",
    },
  },
}
`;

exports[`sets the question state if the answer is incorrect 1`] = `
Object {
  "questData": Object {
    "questions": Object {
      "quest0": Array [
        Object {
          "answer": "8",
          "answerDesc": "Потому что Плутон больше не планета",
          "desc": "На этом месте будет описание ",
          "media": Object {
            "type": "image",
            "uri": "https://",
          },
          "questionText": "Сколько планет в солнечной системе?",
        },
        Object {
          "answer": "24",
          "answerDesc": "Хотелось бы больше, конечно",
          "desc": "На этом месте будет описание вопроса",
          "media": Object {
            "type": "image",
            "uri": "https://",
          },
          "questionText": "Сколько часов в сутках?",
        },
      ],
    },
    "quests": Array [
      Object {
        "cover": "http://...",
        "desc": "Description",
        "id": 0,
        "title": "Test",
      },
      Object {
        "cover": "http://...",
        "desc": "Description",
        "id": 0,
        "title": "Test",
      },
    ],
  },
  "questProgress": Object {
    "quest0": Object {
      "currentAnswer": "15",
      "currentQuestion": 0,
      "currentQuestionState": "INCORRECT",
      "questState": "IN_PROGRESS",
    },
  },
}
`;

exports[`starts a quest 1`] = `
Object {
  "questData": Object {
    "questions": Object {},
    "quests": Array [],
  },
  "questProgress": Object {
    "quest0": Object {
      "currentAnswer": "",
      "currentQuestion": 0,
      "currentQuestionState": "UNANSWERED",
      "questState": "IN_PROGRESS",
    },
  },
}
`;
